### ISSUE TITLE
Arrow functions: `Cannot read property 'getData' of null at remap`

### ISSUE THREAD

**Issue by iam4x** [babel/babel#1846]
First I want to thanks a lot for BabelJS <3

After updating to the latest babeljs, I have problem for bundling a code with an arrow function, using ES5 function declaration just works:

failing code:

```javascript
export default (flux, locale='en') => {
  const {messages} = require(`data/${locale}`);

  flux
    .getActions('locale')
    .switchLocaleSuccess({locale, messages});

  return flux.getStore('locale').getState();
};
```

working code:

```javascript
export default function (flux, locale='en') {
  const {messages} = require(`data/${locale}`);

  flux
    .getActions('locale')
    .switchLocaleSuccess({locale, messages});

  return flux.getStore('locale').getState();
};
```

Full stack trace:

```
Module build failed: TypeError: /Users/iam4x/savemysmartphone/si-front-pdv/test/utils/inject-lang.js: Cannot read property 'getData' of null
    at remap (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/transformers/internal/shadow-functions.js:54:22)
    at NodePath.ReferencedIdentifier (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/transformers/internal/shadow-functions.js:75:14)
    at NodePath.<anonymous> (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/visitors.js:196:17)
    at NodePath.call (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:61:28)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:94:8)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitMultiple (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:92:16)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:122:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitMultiple (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:92:16)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:122:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitMultiple (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:92:16)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:122:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at NodePath.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/path/context.js:111:26)
    at TraversalContext.visitSingle (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:112:12)
    at TraversalContext.visit (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/context.js:124:19)
    at Function.traverse.node (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:72:17)
    at Object.traverse [as default] (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/traversal/index.js:55:14)
    at PluginPass.transform (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/plugin-pass.js:44:27)
    at File.transform (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/file/index.js:596:12)
    at /Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/pipeline.js:131:19
    at File.wrap (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/file/index.js:610:16)
    at Pipeline.transform (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-core/lib/babel/transformation/pipeline.js:128:17)
    at transpile (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-loader/index.js:12:22)
    at Object.module.exports (/Users/iam4x/savemysmartphone/si-front-pdv/node_modules/babel-loader/index.js:69:12)
```

**Comment by iam4x** [comment]
@sebmck Blazing fast :100:

**Comment by sebmck** [comment]
Fixed as of 5.6.12

**Comment by Avien** [comment]
I dont know if its related to this issue but I have an error on arrow functions lexical scoping "this":
It seems the "this" scope refers to the global window and not to the relevant object

```javascript
export class ToolBarService{

    constructor(){
        this.items = {
            'exportTo': null
        }
    }

    get export(){
        return {
            'onClick' : (handler)=>{
                this.items.exportTo = handler;
            }
        }
    }
}
```
this.items is undefined in the 'onClick' function

Important to say that the transpiled code in the bundle.js does seem to look correct:
```javascript
_createClass(ToolBarService,[{
    key:'export',
    get:function get(){
        var _this = this;
        
        return {
            'onClick' : function onClick(handler){
                _this.items.exportTo = handler;
            }
        }
    }
}]);
```

But it doesn't work
Help please
10x
